src/jdk.httpserver/share/classes/sun/net/httpserver/ExchangeImpl.java

```diff
@@ -207,7 +207,7 @@ public void sendResponseHeaders (int rCode, long contentLen)
         }
         this.rcode = rCode;
         String statusLine = "HTTP/1.1 "+rCode+Code.msg(rCode)+"\r\n";
-        OutputStream tmpout = new BufferedOutputStream (ros);
+        ByteArrayOutputStream tmpout = new ByteArrayOutputStream();
```

Comment:
Wouldn't it be better to continue using a BufferedOutputStream here? That would put a limit on how much data would be buffered when writing the headers, in case of long response headers.

Reply:
Forget I asked this. @djelinski pointed out this is necessary to avoid having to flush at line 281. But I see that at line 686 you replaced rawout with a BufferedOutputStream - so do we still need the temporary ByteArrayInputStream? Can't we just write directly to `ros` (possibly assert ros instanceof BufferedOutputStream if we want to make sure...)

Reply:
Before doing that, you might want to check the HTTPS case. As far as I can tell, we don't do any buffering when SSL is in use.

Reply:
I think it was safer for future enhancements the way it was done. The other advantage of using a BAOS over the BOS is that it has better support for underlying hardware that has large frames (e.g. > 64k) to limit the number of OS writes.

Reply:
OK - I was just concerned about the additional buffering this implies. Previously we would only buffer as much as the size of the internal BufferedOutputStream buffer. Now we're buffereing the whole set of headers until the whole set as been encoded.